DAQ_FW/src/main.cpp

```diff
@@ -1,23 +1,23 @@
-
 #include "Logger.h"
 #include "IADCSensor.h"
 
 // object declarations can't be done in setup()
 
-// params: sensorname, sensorID, adcID, gainmode
-IADCSensor SensorTest("TestSensor", 1, 1, GAIN_TWO);
+// params: sensorname, sensorID, adcID
+IADCSensor SensorTest("TestSensor", 1, 0);
```

Comment:
From my understanding, adcID is used for logging to identify which adc a sensor is using. The part of the code that sets this value and the part that actually sets which physical adc to use (which is kind of already determined by the  PCB hw) should be coupled in some way. 

Right now, they are independent from each other so I could pass in an adcID value of 1 while the class is actually using adc0. Then adcID becomes a misrepresentation of the actual hardware creating a potential red herring when debugging.

Reply:
Actually, the i2caddress we use in the class already tells us which physical adc we're using. If you want to make it more readable for the user, you could use a function to map the long hex address like 0xfff213 to a reasonable identifier on the board. For example, the component designator of the ADC (found in the schematic)

Reply:
Can you explain this in more detail, please? From my understanding, the i2c address of the adc is only an 8-bit value based on this:

![image](https://github.com/sfuphantom/DAQ/assets/84810562/48c1eba5-1372-425b-8a86-e9d23cb6b6e6)
 The `ADS1X15_ADDRESS` variable we call in `initialize()` is 0x48 (1001000)
I can create a `#define` in `sys_config` with these to use as the adcID, and use that variable to initialize the adc.  

If not, can you explain where to long hex address comes from, as well as where to look for the ADC component designator, as I had a hard time finding it in the schematic (https://www.ti.com/lit/ds/symlink/ads1115.pdf)

Reply:
An ADC has multiple potential addresses based on how the pins are hooked up as shown in your image. You'll need to talk to Qusai about getting access to the Altium schematics for the DAQ to see where the ADDR pin is connected. The component designator will be in the schematic (and also printed on the PCB itself). Also I overestimated the number of bits so the 8-bit values you are referring to are correct.

This is sort of what I had in mind in terms of the class design https://onlinegdb.com/TTn0xFjND

Three main things going on here:

1 (Line 6):
ADC enum class is defined. This is basically a super strictly typed enum. There is no implicit conversion to int. If you want to get the integer value you have to do an explicit cast like this `static_cast<int>(x)` where `ADCAddress x;` 

2 (Line 24):
Virtual abstract method for obtaining the address. The return value is implemented by the child class and the parent class calls address() to get the value to print to the console. Because the address function **must** return type `ADCAddress`, you won't be able to return a value that isn't U1 or U2. Now the child classes only have to think in terms of component designators instead of individual byte addresses. 

3 (Line 11):
macro for logging. There isn't really a good way to convert to a string from enum, so a simple ternary statement helps to return the correct string literal

A really valuable part of c++ is making the compiler work for you. On line 36 try returning a random integer. You'll get a less cryptic compile time error message with immediate feedback than if you had just passed the i2c address to the constructor at runtime. In that case, it would compile, execute the code and then only when we try to communicate with the physical ADC would we debug and find that the very non obvious issue was our invalid address.